File: 171_LFUCache.cpp
/*
Problem: LFU Cache
Description: Implement Least Frequently Used cache with get and put operations.

Approach: Use hashmap for key->node and frequency->list. Track minFreq. Update frequencies on access.
*/

#include <bits/stdc++.h>
using namespace std;

class LFUCache {
private:
    int capacity;
    int minFreq;
    unordered_map<int, pair<int, int>> keyValue;  // key -> {value, freq}
    unordered_map<int, list<int>> freqList;       // freq -> list of keys
    unordered_map<int, list<int>::iterator> keyIter; // key -> iterator
    
    void updateFreq(int key) {
        int freq = keyValue[key].second;
        freqList[freq].erase(keyIter[key]);
        
        if (freqList[freq].empty()) {
            freqList.erase(freq);
            if (minFreq == freq) minFreq++;
        }
        
        freq++;
        keyValue[key].second = freq;
        freqList[freq].push_front(key);
        keyIter[key] = freqList[freq].begin();
    }
    
public:
    LFUCache(int capacity) : capacity(capacity), minFreq(0) {}
    
    int get(int key) {
        if (keyValue.find(key) == keyValue.end()) return -1;
        
        updateFreq(key);
        return keyValue[key].first;
    }
    
    void put(int key, int value) {
        if (capacity == 0) return;
        
        if (keyValue.find(key) != keyValue.end()) {
            keyValue[key].first = value;
            updateFreq(key);
            return;
        }
        
        if (keyValue.size() >= capacity) {
            int oldKey = freqList[minFreq].back();
            freqList[minFreq].pop_back();
            keyValue.erase(oldKey);
            keyIter.erase(oldKey);
        }
        
        keyValue[key] = {value, 1};
        freqList[1].push_front(key);
        keyIter[key] = freqList[1].begin();
        minFreq = 1;
    }
};

int main() {
    LFUCache cache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    cout << cache.get(1) << endl;  // Output: 1
    cache.put(3, 3);
    cout << cache.get(2) << endl;  // Output: -1 (evicted)
    
    return 0;
}

/*
Time Complexity: O(1) for get and put
Space Complexity: O(capacity)
*/

File: 171_LFUCache.py
"""
Problem: LFU Cache
Description: Implement Least Frequently Used cache with get and put operations.

Approach: Use hashmap for key->node and frequency->list. Track minFreq. Update frequencies on access.
"""

from collections import defaultdict, OrderedDict

class LFUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.minFreq = 0
        self.keyValue = {}  # key -> (value, freq)
        self.freqList = defaultdict(OrderedDict)  # freq -> ordered dict of keys
    
    def _updateFreq(self, key: int) -> None:
        value, freq = self.keyValue[key]
        del self.freqList[freq][key]
        
        if not self.freqList[freq] and self.minFreq == freq:
            self.minFreq += 1
        
        freq += 1
        self.keyValue[key] = (value, freq)
        self.freqList[freq][key] = None
    
    def get(self, key: int) -> int:
        if key not in self.keyValue:
            return -1
        
        self._updateFreq(key)
        return self.keyValue[key][0]
    
    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
        
        if key in self.keyValue:
            self.keyValue[key] = (value, self.keyValue[key][1])
            self._updateFreq(key)
            return
        
        if len(self.keyValue) >= self.capacity:
            oldKey, _ = self.freqList[self.minFreq].popitem(last=False)
            del self.keyValue[oldKey]
        
        self.keyValue[key] = (value, 1)
        self.freqList[1][key] = None
        self.minFreq = 1

def main():
    cache = LFUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print(cache.get(1))  # Output: 1
    cache.put(3, 3)
    print(cache.get(2))  # Output: -1 (evicted)

if __name__ == "__main__":
    main()

"""
Time Complexity: O(1) for get and put
Space Complexity: O(capacity)
"""
